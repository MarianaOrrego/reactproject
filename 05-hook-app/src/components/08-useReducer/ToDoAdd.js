import { useForm } from "../../hooks/useForm";

export const ToDoAdd = ({ handleAddToDo }) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {

        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }
        const nuevaTarea = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddToDo( nuevaTarea );
        reset();
    }

    return (
        <>
            <h4>Agregar tarea</h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input
                    autoComplete='off'
                    className='form-control'
                    name='description'
                    onChange={handleInputChange}
                    placeholder='Tarea pendiente...'
                    type='text'
                    value={description}
                />

                <div className='d-flex justify-content-center mt-3'>
                    <button className='btn btn-outline-primary' type='submit'>
                        Agregar
                    </button>
                </div>

            </form>
        </>
    )
}

