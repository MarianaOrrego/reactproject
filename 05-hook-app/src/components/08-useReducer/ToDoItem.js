export const ToDoItem = ({ tarea, i, handleDelete, handleToggle}) => {
    return (
        <li
            key={tarea.id}
            className='list-group-item'
        >
            <p
                className={` ${tarea.done && 'complete'} `}
                onClick={() => handleToggle(tarea.id)}
            >
                {i + 1}.  {tarea.desc}
            </p>

            <button
                className='btn btn-danger'
                onClick={() => handleDelete(tarea.id)}
            >
                Borrar
            </button>
        </li>
    )
}
