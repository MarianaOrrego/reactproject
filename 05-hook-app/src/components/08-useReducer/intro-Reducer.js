const initialState = [{
    id: 1, 
    tarea: 'Comprar chocolate',
    done: false
}];

const tareaReducer = ( state = initialState, action ) => {

    if (action?.type === 'agregar'){
        return [ ...state, action.payload];
    }

    return state;
}

let tarea = tareaReducer();

const nuevaTarea = {
    id: 2, 
    tarea: 'Comprar leche',
    done: false
}

const agregarAction ={
    type: 'agregar',
    payload: nuevaTarea
}

tarea = tareaReducer( tarea, agregarAction)

console.log( tarea )