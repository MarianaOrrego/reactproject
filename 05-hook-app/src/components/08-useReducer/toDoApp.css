body {
    padding: 50px;
}

h1 {
    text-align: center;
}

h4 {
    text-align: center;
}

p{
    cursor: pointer;
    margin: 0px;
}

.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
}

.complete{
    text-decoration: line-through red;
}