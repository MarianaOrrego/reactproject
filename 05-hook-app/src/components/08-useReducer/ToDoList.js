import PropTypes from 'prop-types';
import { ToDoItem } from "./ToDoItem"

export const ToDoList = ( { tareas, handleDelete, handleToggle } ) => {
    return (
        <ul className='list-group list-group-flush'>
            {
                tareas.map((tarea, i) => (
                    <ToDoItem
                        key={ tarea.id } 
                        tarea= { tarea } 
                        i = { i }
                        handleDelete = { handleDelete } 
                        handleToggle = { handleToggle }
                    />
                ))
            }
        </ul>
    )
}

ToDoList.propTypes = {
    tareas: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}