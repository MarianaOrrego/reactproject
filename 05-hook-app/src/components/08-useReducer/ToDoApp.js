import { useEffect, useReducer } from 'react';
import { toDoReducer } from './toDoReducer';
import './toDoApp.css';
import { ToDoList } from './ToDoList';
import { ToDoAdd } from './ToDoAdd';

const init = () => {

  return JSON.parse(localStorage.getItem('Tareas')) || [];

}

export const ToDoApp = () => {

  const [tareas, dispatch] = useReducer(toDoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('Tareas', JSON.stringify(tareas));
  }, [tareas]);

  const handleDelete = (tareaId) => {

    const action = {
      type: 'delete',
      payload: tareaId
    }

    dispatch(action);
  }

  const handleToggle = (tareaId) => {

    dispatch({
      type: 'toggle',
      payload: tareaId
    })

  }

  const handleAddToDo = ( nuevaTarea ) =>{

    dispatch({
      type: 'add',
      payload: nuevaTarea
    })

  }


  return (
    <div>
      <h1>To do list App ( { tareas.length } )</h1>
      <hr />

      <div className='row'>
        <div className='col-md-6'>

          <ToDoList 
            tareas = { tareas }
            handleDelete = { handleDelete }
            handleToggle = { handleToggle }
          />

        </div>
        <div className='col-md-6'>

          <ToDoAdd 
            handleAddToDo = { handleAddToDo }
          />

        </div>
      </div>

    </div>
  )
}

