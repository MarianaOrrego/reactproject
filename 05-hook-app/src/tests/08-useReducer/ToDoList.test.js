import { shallow } from 'enzyme';
import { ToDoItem } from '../../components/08-useReducer/ToDoItem';
import { demoToDo } from '../fixtures/demoToDo';

describe('Pruebas en <toDoList />', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(
        <ToDoItem
                    tarea= { demoToDo } 
                    handleDelete = { handleDelete } 
                    handleToggle = { handleToggle }
                />          
    )

    test('debe mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();
        
    });
    
    test('debe tener dos <ToDoItem />', () => {

        // expect(wrapper.find('ToDoItem').length).toBe( demoToDo.length );
        // console.log(wrapper.find('ToDoItem').at(0).prop('handleDelete')).toEqual( expect.any(Function));
      
    });
    
})
