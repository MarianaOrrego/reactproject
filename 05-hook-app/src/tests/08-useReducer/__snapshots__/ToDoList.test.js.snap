// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <toDoList /> debe mostrarse correctamente 1`] = `
<li
  className="list-group-item"
>
  <p
    className=" undefined "
    onClick={[Function]}
  >
    NaN
    .  
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
